src/p2p/swarm.rs

```diff
@@ -1,22 +1,104 @@
 use crate::subscription::{SubscriptionFuture, SubscriptionRegistry};
+use anyhow::anyhow;
 use core::task::{Context, Poll};
 use libp2p::core::{
     connection::ConnectionId, multiaddr::Protocol, ConnectedPoint, Multiaddr, PeerId,
 };
 use libp2p::swarm::protocols_handler::{
     DummyProtocolsHandler, IntoProtocolsHandler, ProtocolsHandler,
 };
-use libp2p::swarm::{self, DialPeerCondition, NetworkBehaviour, PollParameters, Swarm};
+use libp2p::swarm::{self, NetworkBehaviour, PollParameters, Swarm};
 use std::collections::{HashMap, HashSet, VecDeque};
+use std::convert::TryFrom;
 use std::time::Duration;
+use std::{fmt, str::FromStr};
+
+/// A wrapper for `Multiaddr` that does **not** contain `Protocol::P2p`.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct MultiaddrWoPeerId(Multiaddr);
```

Comment:
I think we should not abbreviate here so I'd vote for expanding the "Wo" => "Without". I as a non-native speaker am having doubts using "w/o" ever but capitalized and without the slash it seems quite strange.